lib/state/data/middleware.ts

```diff
@@ -74,14 +74,24 @@ export const middleware: S.Middleware = (store) => (
     case 'RESTORE_NOTE_REVISION': {
       const revision = state.data.noteRevisions
         .get(action.noteId)
-        .get(action.version);
+        ?.get(action.version);
 
-      return revision
+      const result = revision
         ? next({
             ...action,
             note: revision,
           })
         : next(action);
+
+      // Restore deleted tags of current revision
+      const tagsToRestore = (revision?.tags ?? [])
+        .map(tagHashOf)
+        .filter((tagHash) => !state.data.tags.has(tagHash));
+      if (tagsToRestore.length > 0) {
+        store.dispatch({ type: 'RESTORE_TAGS', tags: tagsToRestore });
+      }
```

Comment:
we shouldn't need to create a new action for this as we can just as easily watch for this `RESTORE_NOTE_REVISION` action in the tag reducer. currently we're performing tag logic in this place which I don't think is where I'd look for tag changes, plus we're introducing multiple dispatches when one would do.

put another way we've created two places to do things that belong together, so we can just move all of it to the existing tag reducer and avoid creating the new action type.

Reply:
My first approach was to create a new action just in case we might need it in other places but I totally agree that it's not really necessary because it belongs to notes restoration. I'm going to move the tags restoration logic to tags reducer, thanks for pointing me this out, now the code looks cleaner 🙇 .

I've pushed this change in this [commit](https://github.com/Automattic/simplenote-electron/pull/2817/commits/728e23fbbb743c472de0118744014d4281d97e47).

Reply:
## Tag list calculation on middleware

Based on this [comment](https://github.com/Automattic/simplenote-electron/pull/2817#issuecomment-815572261), the tag list calculation is now a bit more complex since it also requires to modify the note tags.

I thought about doing it in the notes reducer, but `combineReducers` only exposes the notes slice on that reducer. One option would be to create a cross slice reducer that receives both states (notes + tags) or even the full one for this action. 

Another alternative I was thinking is to calculate this in the action creator by getting the state (something similar to a thunk action).

I checked the contributing guide but doesn't cover this case, @dmsnell I'd appreciate you feedback on this 🙇 .

## Rules for calculating the tag list

When restoring a revision, the tag list has to be calculated having the following rules in mind:
1. New email tags shouldn't be included because they might grant undesired users permission.
2. Email tags of current note should be kept to prevent modify user permissions, therefore all of them have to be included.
3. Depending on user selection, tags that were previously deleted could be included.

## `APPLY_NOTE_REVISION` action

[`RESTORE_NOTE_REVISION` has the `note` argument](https://github.com/Automattic/simplenote-electron/blob/54b2428867b29ac2e4898a55dc66c92e49ce7cdb/lib/state/action-types.ts#L219-L222) that although it's mandatory, it's not being used when creating the action but populated in the middleware. This generates a Typescript error on the code that [dispatches this action from `revision-selector`](https://github.com/Automattic/simplenote-electron/blob/54b2428867b29ac2e4898a55dc66c92e49ce7cdb/lib/revision-selector/index.tsx#L176-L181).

For this reason, I created a new action `APPLY_NOTE_REVISION` that is dispatched in the middleware after processing `RESTORE_NOTE_REVISION`, which is not dispatched.

Reply:
I created a couple of PRs to explain further the alternatives I mentioned regarding the tag list calculation on the data middleware.

>I thought about doing it in the notes reducer, but combineReducers only exposes the notes slice on that reducer. One option would be to create a cross slice reducer that receives both states (notes + tags) or even the full one for this action.

- [Add cross slice reducer PR](https://github.com/Automattic/simplenote-electron/pull/2838)

>Another alternative I was thinking is to calculate this in the action creator by getting the state (something similar to a thunk action).

- [Use action generator](https://github.com/Automattic/simplenote-electron/pull/2837)

Reply:
> just in case we might need it in other places

I want to reflect on this and bring it back to the heart of what the actions are intended to represent: _something that the customer wants to do_. Unless we were to open up restoring deleted tags there's not "in case" for this, because the action is restoring a note and the fact that we're restoring tags is a _reaction_ to that intention.

> the tag list calculation is now a bit more complex since it also requires to modify the note tags.

we might be conflating vocabulary. I think you are talking about the tag list at the bottom of the note but there's also a major layout piece of the app called "the tag list" in the sidebar. can you explain more where you're having to modify the note tags? what about the `Note` revision passed in as the revision through `OwnProps`? aren't the tags already on that object?

> I thought about doing it in the notes reducer, but combineReducers only exposes the notes slice on that reducer. 

This is a good thing and central to the design of Redux. Each data type operates independently from each other. It's a quick trip to a nightmare when we start spreading around the governance of data outside of itself, vs. being able to jump to the account's tags and ask, "how can these possibly be altered?"

> I checked the contributing guide but doesn't cover this case

the reason it's not covered directly is that we want to avoid the kind of patterns that break out of the core mental model. I can look into making that clearer in the document; thank you for reading it! 😄

there may not actually be much of a conundrum or reason to reach for cross-slice reducers or action thunks, as _the data itself_ doesn't imply the kinds of dependencies these code structures would create. to be clear I'm not opposed to having reducers that stem from multiple inputs, but so far I have tried to model state after that inherent dependency vs. passing multiple bits of state into the reducer (which if we _were_ to do that we'd want to pass the previous state in and not the partially-updated state). see the `ghosts` reducer, for instance, which maintains two separate but dependent values, which by construction will always update atomically together (avoiding data races and consistency issues).

> it's not being used when creating the action but populated in the middleware. This generates a Typescript error

that's a bug in the code that we could fix. IIRC that's from early in the rewrite when we were still deciding on the style we'd use for modifying actions and I probably took a pragmatic decision to introduce the type error and lean on the middleware vs. creating a separate action. at this point I'd say we look into following the example of `CREATE_NOTE` and `CREATE_NOTE_WITH_ID`. we could create something like `RESTORE_NOTE_REVISION_WITH_ID` and humorously convert the other direction.

what we may find, however, could be simpler. with the changes in this PR it might be clearer that in our case we truly are tracking the revision in the UI and basing our decisions on that happen-to-be instance, particularly with tags since those deleted tags no longer exist in state. what we might end up doing can skirt all of the concerns just by passing `note`, which we apparently need to have in order to view the revision in the first place. we generally try to avoid passing _objects_ around when we can pass _ids_ or _references_ instead, but revisions are definitely a different use case and require a bit more ad-hoc data. can you try passing `note` to `restoreRevision` and see if that makes all these questions disappear?

---

thanks for exploring in those two PRs. I wouldn't say I'm thrilled about either of those approaches and so I'd like to keep our focus here and on the heart of the problem: our UI has changed the needs on the data we have previously been passing around - let's match models so the friction is gone.

this is really hard stuff and you're tackling it right away in this repo, bravo. not sure if you have much experience with Redux + React but you're doing a fine job. thank you _very_ much for your effort.

Reply:
>I want to reflect on this and bring it back to the heart of what the actions are intended to represent: something that the customer wants to do. Unless we were to open up restoring deleted tags there's not "in case" for this, because the action is restoring a note and the fact that we're restoring tags is a reaction to that intention.

Ok, makes sense 👍 .

>we might be conflating vocabulary. I think you are talking about the tag list at the bottom of the note but there's also a major layout piece of the app called "the tag list" in the sidebar. can you explain more where you're having to modify the note tags? what about the Note revision passed in as the revision through OwnProps? aren't the tags already on that object?

Yeah sorry for the misleading, when I referenced "the tags list", I meant the tags of the note's revision to restore.

The problem is that we can't pass the revision object as it's in the state when restoring a revision, the reason is that some of the tags, like the email ones, shouldn't be restored because it might add undesired collaborators (this problem was discussed internally p1617789042044500-slack-C036Y8QL4).

>This is a good thing and central to the design of Redux. Each data type operates independently from each other. It's a quick trip to a nightmare when we start spreading around the governance of data outside of itself, vs. being able to jump to the account's tags and ask, "how can these possibly be altered?"

I agree and actually I like the approach of maintaining the reducers totally isolated from the rest. I have experience from previous projects where I used Redux and I can truly recall the nightmare that can turn out if the Redux structure (Actions/Middleware/Reducers) is not properly defined and enforced.

>the reason it's not covered directly is that we want to avoid the kind of patterns that break out of the core mental model. I can look into making that clearer in the document; thank you for reading it! 😄

No problem 😄 ! I think it would be awesome if we cover this case, at least exposing the pattern we want to follow when we need to trigger an action that requires different slices from the state.

>there may not actually be much of a conundrum or reason to reach for cross-slice reducers or action thunks, as the data itself doesn't imply the kinds of dependencies these code structures would create. to be clear I'm not opposed to having reducers that stem from multiple inputs, but so far I have tried to model state after that inherent dependency vs. passing multiple bits of state into the reducer (which if we were to do that we'd want to pass the previous state in and not the partially-updated state). see the ghosts reducer, for instance, which maintains two separate but dependent values, which by construction will always update atomically together (avoiding data races and consistency issues).

One of the powerful features of Redux is that it's quite open so it can be used in different ways. Definitely there're common patterns and good practices but the project's architecture should be the one that drives the use of Redux, for this case makes sense to use it the way you described.

>that's a bug in the code that we could fix. IIRC that's from early in the rewrite when we were still deciding on the style we'd use for modifying actions and I probably took a pragmatic decision to introduce the type error and lean on the middleware vs. creating a separate action. at this point I'd say we look into following the example of CREATE_NOTE and CREATE_NOTE_WITH_ID. we could create something like RESTORE_NOTE_REVISION_WITH_ID and humorously convert the other direction.

Yep, I thought about doing something similar to note creation logic. A new action like what you mentioned (`RESTORE_NOTE_REVISION_WITH_ID`) would work for this case, however I decided to continue with the option described at the bottom of the comment, that doesn't require to create this new action ⬇️ .

>what we may find, however, could be simpler. with the changes in this PR it might be clearer that in our case we truly are tracking the revision in the UI and basing our decisions on that happen-to-be instance, particularly with tags since those deleted tags no longer exist in state. what we might end up doing can skirt all of the concerns just by passing note, which we apparently need to have in order to view the revision in the first place. we generally try to avoid passing objects around when we can pass ids or references instead, but revisions are definitely a different use case and require a bit more ad-hoc data. can you try passing note to restoreRevision and see if that makes all these questions disappear?

No problem, for this option we would need to build the revision object before dispatching the object.

After exploring the different options, I think that in the end the solution is related to where it makes more sense to handle the revision object build. Here I described briefly the options I see:

1. Reducer: This would require a cross slice reducers because we need data from different slices. As commented earlier, this is discarded in favor of keeping the reducers operating independently from each other.
2. Middleware: This is where we're currently handling the logic when we need to transform data upon action execution, as an example we have the create note action. I think we could keep using it for most cases, nevertheless following your comment I tried the fourth option described below.
3. Action creators: Unfortunately we don't have access to the state when creating an action, however this could be solved by adding thunk actions. From my experience, this looks like the most common solution but it's true that it might not fit all projects like this one and besides it comes with its tradeoffs.
4. Selectors: This would be used in the UI before dispatching an action, in this case I think this seems to be the best option since we need to build the revision object depending on UI like including the deleted tags.

>thanks for exploring in those two PRs. I wouldn't say I'm thrilled about either of those approaches and so I'd like to keep our focus here and on the heart of the problem: our UI has changed the needs on the data we have previously been passing around - let's match models so the friction is gone.
>this is really hard stuff and you're tackling it right away in this repo, bravo. not sure if you have much experience with Redux + React but you're doing a fine job. thank you very much for your effort.

Thanks 😊 ! Thank you very much for all the feedback!

I used Redux on some React Native projects in the past and I know that it can be tricky. Redux is a common discussion topic and it usually raises discussions about how to use it, but I always found them very interesting and useful. From my experience one of the most important things regarding this topic is to have as clear as possible architecture, otherwise it's quite easy to lose control.

## Conclusion
Following your comments, I decided to try to pass the note in the `RESTORE_NOTE_REVISION` action in this [commit](https://github.com/Automattic/simplenote-electron/pull/2817/commits/55a321d4673b25fa7fcbbc3603a02cbfaa4055af). For this option, I created a new selector that creates the revision object and it's used in the `revision-selector` component. This way when we dispatch the restore revision note action we pass the final object to be included straight to the state.